rfd/0167-console-service.md

```diff
@@ -0,0 +1,194 @@
+---
+authors: Gabriel Corado (gabriel.oliveira@goteleport.com)
+state: draft
+---
+
+# RFD 0167 - Console service
+
+## Required Approvals
+
+* Engineering: @r0mant && @greedy52
+* Security: @reedloden || @jentfoo
+
+## What
+This RFD proposes a mechanism for Teleport administrators to switch log-level
+and enable profile consumption without restarting the instance.
+
+## Why
+During incidents in production environments, debug logs and diagnosis profiles
+(such as CPU and Memory allocation) are often unavailable for timely
+troubleshooting. To enable those, administrators need to restart the instance,
+potentially disrupting the application state and obscuring the issue they're
+trying to solve.
+
+## Details
+
+A separate service will be introduced specifically for troubleshooting
+purposes. This will ensure the current diagnosis service is kept as is, not
+breaking current integrations or introducing behavior changes. The new service
+will also always be available (contrasting with the diagnosis service, which
+is optional and disabled by default), so in scenarios where users need to use
+it, they don't need to restart Teleport.
+
+In addition, the new service will listen into a Unix socket instead of TCP. This
+will make discoverability easier and discourage external usage, as the API is
+designed for internal usage.
+
+The `pprof` endpoints on diagnostics will still be available for existent
+integrations, such as the usage of
+`go tool pprof http://diag-addr/debug/pprof/profile`.
+
+### New service
+
+Teleport will start listening using a Unix socket located at
+`/tmp/teleport-diag-<node-id>.sock`.
```

Comment:
Is there an argument for using `/tmp` rather than `/var/run` or similar ? I *believe* this is the more idiomatic location for sockets.

Reply:
Nothing specific about that. I can update it to be on `/var/run`.

Reply:
@strideynet After testing, I noticed that placing the socket at `/var/run` will require Teleport to run as a root user. Otherwise, it will fail to start due to a lack of permission. Placing it at `/tmp` would be better to avoid blocking existing deployments (as it doesn't require additional permission).

Note: The `teleport console` commands might still require `sudo` to execute, either due to lack of user access to the `/tmp` directory, access to the configuration file (by default placed at `/etc/teleport.yaml`, or `data_dir` directory (I'll add a friendly message so users don't get stuck).

Reply:
This thread caught my attention as I'm working on a daemon for VNet at the moment, though it will always run as root.

FWIW, there's also `XDG_RUNTIME_DIRECTORY` on compatible systems for unprivileged users (I found it in [systemd docs](https://systemd.io/TEMPORARY_DIRECTORIES/)). Though storing the socket in the data directory does seem appealing in the context of this RFD.